scripts/backtest_mm_example.py

```diff
@@ -0,0 +1,181 @@
+            pass
+        else:
+            df = self.candle.candles_df
+            df['ask_price'] = df["open"] * (1 + self.ask_spread_bps / 10000)
+            df['bid_price'] = df["open"] * (1 - self.bid_spread_bps / 10000)
+            df['buy_amount'] = df['low'].le(df['bid_price']) * self.order_amount
+            df['sell_amount'] = df['high'].ge(df['ask_price']) * self.order_amount
+            df['fees_paid'] = (df['buy_amount'] * df['bid_price'] + df['sell_amount'] * df['ask_price']) * self.fee_bps / 10000
+            df['base_delta'] = df['buy_amount'] - df['sell_amount']
+            df['quote_delta'] = df['sell_amount'] * df['ask_price'] - df['buy_amount'] * df['bid_price'] - df['fees_paid']
+
+        if self.candle.is_ready and self.results_df is None:
+            df.to_csv(self.csv_path, index=False)
+            self.results_df = df
+            msg = "Backtesting complete - run 'status' to see results."
+            self.log_with_clock(logging.INFO, msg)
+            self.notify_hb_app_with_timestamp(msg)
+
+    def on_stop(self):
+        self.candle.stop()
+
+    def get_trades_df(self, df):
+        total_buy_trades = df['buy_amount'].ne(0).sum()
+        total_sell_trades = df['sell_amount'].ne(0).sum()
+        amount_bought = df['buy_amount'].sum()
+        amount_sold = df['sell_amount'].sum()
+        end_price = df.tail(1)['close'].values[0]
+        amount_bought_quote = amount_bought * end_price
+        amount_sold_quote = amount_sold * end_price
+        avg_buy_price = np.dot(df['bid_price'], df['buy_amount']) / amount_bought
+        avg_sell_price = np.dot(df['ask_price'], df['sell_amount']) / amount_sold
+        avg_total_price = (avg_buy_price * amount_bought + avg_sell_price * amount_sold) / (amount_bought + amount_sold)
+
+        trades_columns = ["", "buy", "sell", "total"]
+        trades_data = [
+            [f"{'Number of trades':<27}", total_buy_trades, total_sell_trades, total_buy_trades + total_sell_trades],
+            [f"{f'Total trade volume ({self.base})':<27}",
+             round(amount_bought, self.precision),
+             round(amount_sold, self.precision),
+             round(amount_bought + amount_sold, self.precision)],
+            [f"{f'Total trade volume ({self.quote})':<27}",
+             round(amount_bought_quote, self.precision),
+             round(amount_sold_quote, self.precision),
+             round(amount_bought_quote + amount_sold_quote, self.precision)],
+            [f"{'Avg price':<27}",
+             round(avg_buy_price, self.precision),
+             round(avg_sell_price, self.precision),
+             round(avg_total_price, self.precision)],
+        ]
+        return pd.DataFrame(data=trades_data, columns=trades_columns)
+
+    def get_assets_df(self, df):
+        for connector_name, connector in self.connectors.items():
```

Comment:
What do you think about extracting 
```python
        for connector_name, connector in self.connectors.items():
            base_bal_start = float(connector.get_balance(self.base))
            quote_bal_start = float(connector.get_balance(self.quote))
        base_bal_change = df['base_delta'].sum()
        quote_bal_change = df['quote_delta'].sum()
        base_bal_end = base_bal_start + base_bal_change
        quote_bal_end = quote_bal_start + quote_bal_change
        start_price = df.head(1)['open'].values[0]
        end_price = df.tail(1)['close'].values[0]
```
from here and [get_performance_df](https://github.com/hummingbot/hummingbot/pull/6303/files#diff-2c3a80b605d76ff6cabdad4313b3bb9b4865dd064a3cb46d6fd91336ad62be05R125) to a separate function?

Reply:
We'll merge this in as an example, but feel free to propose improvements to this script.